### PR TITLE
Remove hydration URL check

### LINKED ISSUE #8872: encoded character in url causes "Initial URL does not match URL at time of hydration"
### Reproduction

https://github.com/WillSmithTE/grunge-test-2

https://19przkzqw9.execute-api.us-west-2.amazonaws.com

1. open the app above
2. open the whistler blog
3. infinite refresh because "Initial URL does not match URL at time of hydration"

This is deployed with the grunge stack on lambda, and the whistler blog title (and url) has an encoded character, "," (or "%2C")

in the html, this is the remix context, note the "decoded" comma

```
<script>
            window.__remixContext = {
                "url": "/blog/Good%20day%20at%20Whistler,%20but%20a%20bit%20icy",
                "state": {
                    "loaderData": {
                        "root": {
                            "user": null
                        },
                        "routes/blog.$title": {
                            "blog": {
                                "id": "a",
                                "title": "Good day at Whistler, but a bit icy"
                            }
                        }
                    },
                    "actionData": null,
                    "errors": null
                },
                "future": {
                    "v3_fetcherPersist": false,
                    "v3_relativeSplatPath": false,
                    "v3_throwAbortReason": false
                }
            };
        </script>
        ```

https://github.com/remix-run/remix/assets/26863411/091a8e6d-cd1e-4a07-9727-bcddb5e449be

### System Info

```shell
System:
    OS: macOS 14.1
    CPU: (11) arm64 Apple M3 Pro
    Memory: 130.39 MB / 18.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.11.0 - ~/.nvm/versions/node/v20.11.0/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.4 - ~/.nvm/versions/node/v20.11.0/bin/npm
    Watchman: 2023.12.04.00 - /usr/local/bin/watchman
  Browsers:
    Brave Browser: 121.1.62.153
    Chrome: 121.0.6167.184
    Safari: 17.1
  npmPackages:
    @remix-run/architect: ^2.7.2 => 2.7.2 
    @remix-run/css-bundle: ^2.7.2 => 2.7.2 
    @remix-run/dev: ^2.7.2 => 2.7.2 
    @remix-run/node: ^2.7.2 => 2.7.2 
    @remix-run/react: ^2.7.2 => 2.7.2 
    vite: ^5.0.0 => 5.1.4
```

### Used Package Manager

npm

### Expected Behavior

Page loads normally

### Actual Behavior

Page infinitely refreshes cause of the mismatch between encoded and decoded comma

### PR DESCRIPTION
**Pull request by brophdawg11**
**tl;dr;** this was only needed in React 17 and can be dropped now that Remix v2 requires React 18.  This will get rid of some false-positive scenarios we have today.

See changeset and https://github.com/remix-run/remix/issues/1757#issuecomment-2294186651 for details.

Closes #8872, #7529

### CODE REVIEW

**Review by jacob-ebey — APPROVED** [review]

**Comment by mikkelwf** [comment]
We´re experiencing this bug in production right now. 
A fast merge and subsequent release would be appreciated.. :)
